js-packages/@prestojs/viewmodel/src/fields/AsyncChoices.ts

```diff
@@ -0,0 +1,242 @@
+ * called from a component or hook that deals with async choices when calling
+ * `list` and `retrieve` respectively. The return value from the hook is passed to
+ * the corresponding function.
+ *
+ * For a default implementation see [AsyncChoices](doc:AsyncChoices)
+ *
+ * @extract-docs
+ * @menu-group Async Choices
+ */
+export interface AsyncChoicesInterface<ItemType, ValueType> {
+    /**
+     * If true then multiple values can be selected. When this is true
+     * retrieve() will be passed and return an array rather than a single value.
+     */
+    multiple: boolean;
+    /**
+     * Hook that returns dependencies for listing choices. This is useful to store state
+     * for things like pagination. The value returned here is passed to `list`.
+     *
+     * What this function is passed depends on the implementation but when used with
+     * [useAsyncChoices](doc:useAsyncChoices) it will be passed `value` (the currently
+     * selected value), `query` (the query object eg. to filter results with) and `listOptions`
+     * (any additional options passed on the `listOptions` prop to `useAsyncChoices`).
+     * Official presto widgets all use `useAsyncChoices` and so will use these parameters.
+     */
+    useListProps(args: any): any;
+    /**
+     * Function to resolve a list of choices based on the provided params.
+     *
+     * What this function is passed depends on the implementation but when used with
+     * [useAsyncChoices](doc:useAsyncChoices) it will be passed `value` (the currently
+     * selected value), `query` (the query object eg. to filter results with), `paginator` (the current
+     * paginator if any) and `listOptions` (any additional options passed on the `listOptions` prop to `useAsyncChoices`).
+     * Official presto widgets all use `useAsyncChoices` and so will use these parameters.
+     */
+    list(params: Record<string, any>): Promise<ItemType[]>;
+
+    /**
+     * Hook that returns dependencies for retrieving an existing value of a choice. This is
+     * useful for things like hooking into an existing cache (eg. [useViewModelCache](doc:useViewModelCache])). The value
+     * returned here is passed as the second parameter to `retrieve`. In addition the `existingValues` key is
+     * passed through to `useAsyncValue` as the list of items it can resolve existing values from.
+     *
+     * What this function is passed depends on the implementation but when used with
+     * [useAsyncChoices](doc:useAsyncChoices) it will be passed `id` if there's a current value
+     * and it's not an array, `ids` if there's a current value and it is an array, `existingValues` which is
+     * the values returned by `list` (may be null if `list` not yet called) and `retrieveOptions`
+     * (any additional options passed on the `retrieveOptions` prop to `useAsyncChoices`).
+     * Official presto widgets all use `useAsyncChoices` and so will use these parameters.
+     */
+    useRetrieveProps(args: any): any;
```

Comment:
Same comment as `useListProps` above

Reply:
Same kind of deal - most obvious usage I had for it was reading from cache always for existing values to avoid unnecessary fetches, eg https://github.com/prestojs/prestojs/pull/67/files#diff-013e735fdb71bd7fe985d673e4eb416cR62